src/test/scala/chiselTests/BoringUtilsSpec.scala

```diff
@@ -125,4 +126,145 @@ class BoringUtilsSpec extends ChiselFlatSpec with ChiselRunners {
+    import chisel3.experimental.hierarchy._
+    @instantiable
+    class Bar extends RawModule {
+      @public val a_wire = WireInit(UInt(1.W), DontCare)
+      dontTouch(a_wire)
+    }
+    class Foo extends RawModule {
+      val a = IO(Output(UInt()))
+
+      val bar = Instance(Definition((new Bar)))
+
+      a := BoringUtils.bore(bar.a_wire)
+    }
+    val e = intercept[Exception] {
+      circt.stage.ChiselStage.emitCHIRRTL(new Foo, args)
+    }
+    e.getMessage should include("Cannot bore across a Definition/Instance boundary")
+  }
+
+  it should "work boring upwards" in {
+    import chisel3.experimental.hierarchy._
+    class Bar(parentData: Data) extends RawModule {
+      val q = Wire(UInt(1.W))
+      q := BoringUtils.bore(parentData)
+    }
+    class Foo extends RawModule {
+      val a = IO(Input(UInt(1.W)))
+      val bar = Module(new Bar(a))
+    }
+    (circt.stage.ChiselStage
+      .emitCHIRRTL(new Foo)
+      .split('\n')
+      .map(_.takeWhile(_ != '@'))
+      .map(_.trim) should contain).inOrder(
+      "module Bar :",
+      "input q_bore : UInt<1>",
+      "q <= q_bore_1", // Do normal connection before secret ones
+      "q_bore_1 <= q_bore",
+      "module Foo :",
+      "input a : UInt<1>",
+      "bar.q_bore <= a"
+    )
+  }
+
+  it should "not affect DataMirror.getPorts, as they are secret (shhh)" in {
+    import chisel3.reflect.DataMirror
+    class Bar extends RawModule {
+      val a_wire = WireInit(UInt(1.W), DontCare)
+      dontTouch(a_wire)
+    }
+    class Foo extends RawModule {
+      val a = IO(Output(UInt()))
+      val bar = Module(new Bar)
+      val preBore = DataMirror.modulePorts(bar)
+      a := BoringUtils.bore(bar.a_wire)
+      val postBore = DataMirror.modulePorts(bar)
+      preBore should be(postBore)
+    }
+    circt.stage.ChiselStage.emitCHIRRTL(new Foo, args)
+  }
+  it should "be ok if with .toDefinition, if it is after boring" in {
+    import chisel3.reflect.DataMirror
+    class Bar extends RawModule {
+      val a_wire = WireInit(UInt(1.W), DontCare)
+      dontTouch(a_wire)
+    }
+    class Foo extends RawModule {
+      val a = IO(Output(UInt()))
+      val bar = Module(new Bar)
+      a := BoringUtils.bore(bar.a_wire)
+      bar.toDefinition
+    }
+
+    circt.stage.ChiselStage.emitCHIRRTL(new Foo, args)
+  }
+  it should "error if ever bored after calling .toDefinition" in {
+    import chisel3.reflect.DataMirror
+    class Bar extends RawModule {
+      val a_wire = WireInit(UInt(1.W), DontCare)
+      dontTouch(a_wire)
+    }
+    class Foo extends RawModule {
+      val a = IO(Output(UInt()))
+      val bar = Module(new Bar)
+      bar.toDefinition
+      a := BoringUtils.bore(bar.a_wire)
+    }
+
+    val e = intercept[Exception] {
+      circt.stage.ChiselStage.emitCHIRRTL(new Foo, args)
+    }
+    e.getMessage should include("Can only bore into modules that have not been converted to a Definition")
+  }
```

Comment:
Nice error test case. This could also be cleaned up by removing the dontTouch and port `a`.